### PR TITLE
[WIP] Redfish Client & Redfish OOB Driver

### PR DESCRIPTION
**Pull request by GabrielBrascher**
Implementation overview:
- Redfish Java client: this PR adds a Java Client for Redfish; thus, making Redfish actions available to the HA workflow via an OOB driver.
- OOB driver: a new Out-of-band driver was created for Redfish, allowing to integrate the Redfish Client with the CloudStack Out-of-band management implementation.

This PR tracks and helps on monitoring and developing the Redfish OOBM implementation

**Note:** this feature was only possible with the help of our friends at [Yourhosting](https://www.yourhosting.nl/).

### CODE REVIEW

**Comment by GabrielBrascher** [comment]
Testing the plugin on CloudStack workflow. The following log shows that RedfishOutOfBandManagementDriver is propperly loaded into the CloudStack.
```
2020-01-08 05:41:06,511 INFO  [o.a.c.s.m.m.i.DefaultModuleDefinitionSet] (main:null) (logid:) Loaded module context [nested-cloudstack] in 109 ms
2020-01-08 05:41:06,511 INFO  [o.a.c.s.m.m.i.DefaultModuleDefinitionSet] (main:null) (logid:) Loading module context [redfish] from URL [jar:file:/usr/share/cloudstack-management/lib/cloudstack-4.14.0.0-SNAPSHOT.jar!/META-INF/cloudstack/redfish/spring-redfish-context.xml]
2020-01-08 05:41:06,511 INFO  [o.a.c.s.m.m.i.DefaultModuleDefinitionSet] (main:null) (logid:) Loading module context [redfish] from URL [jar:file:/usr/share/cloudstack-management/lib/cloudstack-4.14.0.0-SNAPSHOT.jar!/META-INF/cloudstack/outofbandmanagement/spring-core-lifecycle-outofbandmanagement-context-inheritable.xml]
2020-01-08 05:41:06,511 INFO  [o.a.c.s.m.m.i.DefaultModuleDefinitionSet] (main:null) (logid:) Loading module context [redfish] from URL [jar:file:/usr/share/cloudstack-management/lib/cloudstack-4.14.0.0-SNAPSHOT.jar!/META-INF/cloudstack/core/spring-core-lifecycle-core-context-inheritable.xml]
2020-01-08 05:41:06,511 INFO  [o.a.c.s.m.m.i.DefaultModuleDefinitionSet] (main:null) (logid:) Loading module context [redfish] from URL [jar:file:/usr/share/cloudstack-management/lib/cloudstack-4.14.0.0-SNAPSHOT.jar!/META-INF/cloudstack/system/spring-framework-config-system-context-inheritable.xml]
2020-01-08 05:41:06,511 INFO  [o.a.c.s.m.m.i.DefaultModuleDefinitionSet] (main:null) (logid:) Loading module context [redfish] from URL [jar:file:/usr/share/cloudstack-management/lib/cloudstack-4.14.0.0-SNAPSHOT.jar!/META-INF/cloudstack/system/spring-core-system-context-inheritable.xml]
2020-01-08 05:41:06,511 INFO  [o.a.c.s.m.m.i.DefaultModuleDefinitionSet] (main:null) (logid:) Loading module context [redfish] from URL [jar:file:/usr/share/cloudstack-management/lib/cloudstack-4.14.0.0-SNAPSHOT.jar!/META-INF/cloudstack/system/spring-contrail-system-context-inheritable.xml]
2020-01-08 05:41:06,511 INFO  [o.a.c.s.m.m.i.DefaultModuleDefinitionSet] (main:null) (logid:) Loading module context [redfish] from URL [jar:file:/usr/share/cloudstack-management/lib/cloudstack-4.14.0.0-SNAPSHOT.jar!/META-INF/cloudstack/bootstrap/spring-bootstrap-context-inheritable.xml]
2020-01-08 05:41:06,588 DEBUG [o.a.c.f.c.i.ConfigDepotImpl] (main:null) (logid:) Retrieving keys from RedfishOutOfBandManagementDriver
2020-01-08 05:41:06,660 DEBUG [o.a.c.s.l.r.ExtensionRegistry] (main:null) (logid:) Registering extension [REDFISH] in [Out Of Band Management Drivers Registry]
2020-01-08 05:41:06,660 DEBUG [o.a.c.s.l.r.RegistryLifecycle] (main:null) (logid:) Registered org.apache.cloudstack.outofbandmanagement.driver.redfish.RedfishOutOfBandManagementDriver@10683f87
2020-01-08 05:41:06,660 INFO  [o.a.c.s.l.CloudStackExtendedLifeCycle] (main:null) (logid:) Configuring CloudStack Components
2020-01-08 05:41:06,661 INFO  [o.a.c.s.l.CloudStackExtendedLifeCycle] (main:null) (logid:) Done Configuring CloudStack Components
2020-01-08 05:41:06,661 INFO  [o.a.c.s.m.m.i.DefaultModuleDefinitionSet] (main:null) (logid:) Loaded module context [redfish] in 150 ms
2020-01-08 05:41:06,663 INFO  [o.a.c.s.m.m.i.DefaultModuleDefinitionSet] (main:null) (logid:) Starting module [system]```

File: plugins/outofbandmanagement-drivers/redfish/src/main/java/org/apache/cloudstack/outofbandmanagement/driver/redfish/RedfishOutOfBandManagementDriver.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+//    private ExecutorService redfishExecutor = Executors.newFixedThreadPool(OutOfBandManagementService.SyncThreadPoolSize.value(), new NamedThreadFactory("RedfishDriver"));
+    private RedfishWrapper redfishWrapper = new RedfishWrapper();
+
+    //TODO for now they are set to ConfigKey.Scope.Global, need to review it
+    public static final ConfigKey<Boolean> IGNORE_SSL_CERTIFICATE = new ConfigKey<Boolean>("Advanced", Boolean.class, "redfish.ignore.ssl", "false",
+            "Default value is true, ensuring that the client requests validate the certificate when using SSL. If set to true the redfish client will ignore SSL certificate validation when sending requests to a Redfish server.",
+            true, ConfigKey.Scope.Global);
+
+    public static final ConfigKey<Boolean> USE_HTTPS = new ConfigKey<Boolean>("Advanced", Boolean.class, "redfish.use.https", "false",
+            "If set to false the redfish client will ignore SSL certificate validation when sending requests to a Redfish server.", true, ConfigKey.Scope.Global);
```
Comment (wido) [inline plugins/outofbandmanagement-drivers/redfish/src/main/java/org/apache/cloudstack/outofbandmanagement/driver/redfish/RedfishOutOfBandManagementDriver.java]: This description seems odd. I think it was a copy/paste from the previous config key?

I think this one should say something like: "Use HTTPS/SSL for all connections"

File: utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+            builder.setUserInfo(username, password);
+
+            HttpGet httpReq = new HttpGet(builder.build());
+            httpReq.addHeader("accept", "application/json");
+
+            HttpClient client;
+            if (ignoreSSL) {
+                client = ignoreSSLCertValidator();
+            } else {// TODO spawn a regular http client
+                throw new RedfishException("Redfish Client does not support SSL certificates. Please set the client to ignore SSL certificates.");
```
Comment (wido) [inline utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java]: Why is this needed? Because you can just spawn a regular HTTP client without anything else. It will then complain if it can't validate the SSL certificate.

File: utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+            httpReq.addHeader(HTTP.CONTENT_TYPE, "application/json");
+            httpReq.setEntity(new StringEntity(jsonToSend.toString()));
+
+            System.out.println("\nRequest: " + httpReq);
+
+            HttpClient client;
+            if (ignoreSSL) {
+                client = ignoreSSLCertValidator();
+            } else {// TODO spawn a regular http client
+                throw new RedfishException("Redfish Client does not support SSL certificates. Please set the client to ignore SSL certificates.");
```
Comment (wido) [inline utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java]: Same as with executeGetRequest.

Can't these share the same code?

File: utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+        }
+    }
+
+    /**
+     *
+     */
+    private String validateAddressAndPrepareForUrl(String hostAddress) {
+        if (NetUtils.isValidIp6(hostAddress)) {
+            return String.format("[%s]", hostAddress);
+        } else if (NetUtils.isValidIp6(hostAddress)) {
```
Comment (wido) [inline utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java]: This should validate for a valid IPv4 and not IPv6 I think?

File: utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java [hunk 5]
```diff
@@ -0,0 +1,423 @@
...
+            return hostAddress;
+        } else {
+            throw new RedfishException(String.format("Redfish host address '%s' is not a valid IPv4 or IPv6 address", hostAddress));
```
Comment (wido) [inline utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java:273]: This shouldn't be the case always. Because a hostname should also be supported.

People might have a DNS entry pointing to the iDRAC/IPMI of their server(s).

File: utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java [hunk 6]
```diff
@@ -0,0 +1,423 @@
...
+            BufferedReader streamReader = new BufferedReader(new InputStreamReader(in, "UTF-8"));
+
+            // TODO Nasty handling, will be used for testing. We should parse it as JSON.
+            // Check also the Apache Olingo library for manipulating Odata
+            jsonString = streamReader.readLine();
+        } catch (UnsupportedOperationException | IOException e) {
+            throw new RedfishException("Failed to process system Response", e);
+        }
+
+        String open = "/redfish/v1/Systems/";
```
Comment (wido) [inline utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java]: Re-use SYSTEMS_URL_PATH here?

**Review by wido — COMMENTED** [review]
Left a few comments.

Looking good though!

File: utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java (hunk 5 above)
Comment (GabrielBrascher) [reply utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java:273]: The `InternetDomainName.isValid(hostAddress)` validation should be enough to validate a domain name.

I will be soon committing a few updates,  one of them is a new JUnit test that ensures this method to accept a domain name (e.g. mydomain.org).

**Comment by GabrielBrascher** [comment]
Closing Draft PR.

- _Official_ PR is at https://github.com/apache/cloudstack/pull/4175
- new branch is [pcextreme:redfish-oob](https://github.com/PCextreme/cloudstack/tree/redfish-oob)
